embedded-nal-async/src/stack/tcp.rs

```diff
@@ -170,3 +170,25 @@ impl<T: TcpClientStack> TcpClientStack for &mut T {
 		T::close(self, socket)
 	}
 }
+
+/// This trait is implemented by TCP/IP stacks. In contrast to the TcpClientStack trait,
+/// this trait allows creating multiple connections using the same TCP/IP stack.
+///
+/// The trait relies on an associated type, `TcpConnection` which is returned when
+/// creating a TCP connection.
+pub trait TcpClient: embedded_io::Io {
+	/// Type returned when creating a connection.
+	type TcpConnection<'m>: embedded_io::asynch::Read + embedded_io::asynch::Write
+	where
+		Self: 'm;
+
+	/// Future returned by `connect` function.
+	type ConnectFuture<'m>: Future<Output = Result<Self::TcpConnection<'m>, Self::Error>> + 'm
```

Comment:
I don't think that using lifetime `'m` for `Self::TcpConnection<'m>` achieves the advertised goal of "this trait allows creating multiple connections using the same TCP/IP stack".

The problem of `'m` is that it lives as long as the `&'m mut self` reference to `TcpClient` on line 193 below. What this means is that I can keep around the `Self::TcpConnection<'m>` instance *only as long as the `&'m mut self` reference to `TcpClient` is alive*. This results in two problems:
- I can't hold an instance of `Self::TcpConnection<'m>` for a longer period than `'m'` I.e. I can't implement cache and connection re-use, as I do [here](https://github.com/ivmarkov/embedded-svc-impl/blob/master/src/asynch/http/client.rs#L237)
- Worse, I can't really create *multiple* `TcpConnection` instances as currently each created `TcpConnection` essentially holds a reference to `&mut TcpClient`. So I can in fact have - at any point in time - a **single** `TcpConnection`?

The solution IMO is that the `TcpClient` trait must be parameterized by another lifetime (say, `'t`) and the returned `TcpConnection` instances should return reference this (longer) lifetime `'t` and not `'m`, as suggested [here](https://github.com/ivmarkov/embedded-nal/blob/shared-stack/embedded-nal-async/src/stack/tcp.rs#L186). (For STD implementations `'t` is actually `'static` as implemented [here](https://github.com/ivmarkov/embedded-svc-impl/blob/master/src/asynch/stdnal.rs#L35)).

See also [this PR](https://github.com/lulf/embedded-nal/pull/1/files) for all suggested changes.

Reply:
I'm not sure introducing another lifetime helps. The way TcpClient is intended is that you create the clients on application init, and then pass them to whatever needs to create connections.

Having the connection outlive the client means that you need to store connection state somewhere global, but with this trait the connection can rely on tcpclient to be around until it is dropped.

Reply:
> I'm not sure introducing another lifetime helps. The way TcpClient is intended is that you create the clients on application init, and then pass them to whatever needs to create connections.

OK so you are modelling Option "a" from my "philosophical" comment further below. No complaints with that and I see your arguments why we do not necessarily need to model the notion of a "Stack" from embassy as a trait - we might not need that in the embedded world indeed.

In that case:
- How about re-wording the comment above the `TcpClient` trait? It currently says - "... this trait allows creating multiple connections using the same TCP/IP stack.". In retrospective I see what you mean, but that comment might still be misleading as its direct interpretation led me to believe that I'm supposed to be capable of creating multiple connections via a single `TcpClient` instance;
- It might also help if we rename `TcpClient` to `TcpClientSocket`, because `TcpClient` is really a socket in a way. Granted, a one that can also be in disconnected state initially, but that's just like the native smoltcp `TcpSocket` type, isn't it?
- The current design where `connect` takes `&mut self` reference to `TcpClient(Socket)` makes caching `TcpConnection` very difficult. Can you look [here](https://github.com/ivmarkov/embedded-svc-impl/blob/master/src/asynch/http/client.rs#L234) again? Not really sure how to cache the `connection` instance inside `Client`. I would expect my MQTT client to be capable of auto-reconnecting if the socket connection dies - your [esp8266 example](https://github.com/drogue-iot/drogue-device/blob/main/examples/nrf52/microbit/esp8266/mqtt/src/main.rs#L213) does not do that; does this mean that on socket error the whole thing falls apart, or is the actor framework somehow automagically restarting the actor? Ditto for the HTTP client, where it is not just about re-connecting, but in fact about **re-using** the same (in the worst case - TLS) socket across multiple requests. The only way I see this possible to fix is by introducing explicitly an `HttpConnection` trait that will wrap the `TcpConnection` instance by taking a `&mut self` reference to the `Client`, which in turns wraps and owns a `TcpClient`. A bit too heavyweight solution.

Reply:
I agree the comment should be updated. We can also rename it, but I'd be wary about using the name 'Socket', as sockets are not reusable once closed/dropped (much like a connection), and this could set wrong expectations as well. Not sure what a good name would be, maybe TcpConnectionFactory :) IMHO TcpClient is an OK compromise, but maybe some better name exists.

On the connection reuse, I think it's a double edged sword. Having the mqtt and http client wrap the tcp client I think makes them less flexible, but more user friendly.  I need to think/experiment a bit with this to see what can be done, but I think a similar fix to what you propose might be needed. In general your trying to reuse resources across multiple requests, much like TcpClient tried to reuse the tcp stack for multiple connections, so I think the same pattern emerges.

Reply:
> [Here's an alternative design](https://github.com/ivmarkov/embedded-nal/blob/shared-stack/embedded-nal-async/src/stack/tcp.rs#L197) which avoids the lifetime problem at the expense of runtime errors if you try to read/write from a `TcpClientSocket` instance which is not connected yet. But... this might well happen even for a connected socket if the other side did disconnect, isn't it?
> 

That's true, one cannot prevent bad things from happening at run time, the underlying network might be unavailable as well. I'm a bit divided on this though. On one hand, yes the API might make implementation a bit simpler. On the other hand though, making the API simpler and harder to misuse also has some value, since there are generally more users of the API than implementors.

But I think as a generic trait, what you propose is probably a better starting point and more flexible in use.

> As for renaming `TcpClient` to `TcpClientSocket`... I see your concerns but on the other hand, that did not stop the smoltcp folks naming their own thing `TcpSocket` even if it is a bit different from a STD socket.
> 

I think that's unfortunate, but I guess inventing new terms could also be considered confusing. In your alernative trait through, I think it calling it `TcpClientSocket` is more appropriate.

> As for the mqtt and http clients wrapping a tcp client making them less flexible - yes and no. E.g. if you do not want to use the full-fledged HTTP client, you can just re-use the `Request` trait impl, which only needs a `Read+Write` trait impl. The benefit of wrapping the tcp client (socket) is exactly that you get easier management of the socket connection, with automatic reconnect and reuse, which I thought is one of the main reasons to push for the `TcpClient` trait anyway?
> 
> By the way, how I would imagine TLS would be layered is by a `TcpClient(Socket)` impl, which takes another, raw `TcpClient(Socket)` impl and layers on top `embedded-tls` in a way which is transparent for the consumer...

Yeah, I went down this route initially, but it's a hard to make it transparent because you typically want to configure TLS server name, CA certificates, client certs. etc, and they cannot be represented in the TcpClient connect API.

Reply:
(Side topic, but perhaps related, as ideally we should be able to abstract out the TLS machinery as well.)

> > By the way, how I would imagine TLS would be layered is by a `TcpClient(Socket)` impl, which takes another, raw `TcpClient(Socket)` impl and layers on top `embedded-tls` in a way which is transparent for the consumer...
> 
> Yeah, I went down this route initially, but it's a hard to make it transparent because you typically want to configure TLS server name, CA certificates, client certs. etc, and they cannot be represented in the TcpClient connect API.

Huh that's very unfortunate. I do not have in-dept understanding of TLS, but I was assuming, that whatever you configure should be **independent** of the host to which you are connecting, and therefore should not be passed-in to the `connect` method, but earlier, when the `TcpClient(Socket)` is *constructed*. I.e.:
- CA certificates are always the same
- Client certificate should be the same too, regardless of the host?
- Not sure about TLS server name though... :(

Reply:
All of these settings depends on what you're connecting to. Limiting them to be configured at client construction time can be fine for many use cases and is certainly possible. It just means in practice you can only connect to the same host with the same client: this is probably fine, but it is weird since connect() in the trait is passed the SocketAddr which means you _can_ use it to connect to whatever IP + port you want, so shouldn't that also be in the constructor of the type... :) 

TLS also requires some resources only at connect time, like a random number generator. Borrowing it only on connect means you don't need to deal with complex lifetimes and shared access to the rng in the application.

But let's not finalize that design here :D Would it make sense then to update command + replace TcpClient with the TcpClientSocket trait as you propose for this PR?

Reply:
> But let's not finalize that design here :D Would it make sense then to update command + replace TcpClient with the TcpClientSocket trait as you propose for this PR?

Yes. But only if you feel that's the right direction too. Does it make sense to try to implement the new trait for smoltcp and esp8266 AT as you did for the original trait? I'll do the same for async STD and I'll also try to use the new trait in my experimental HTTP client traits.

Reply:
Async STD impl in progress [here](https://github.com/ivmarkov/embedded-svc-impl/blob/master/src/asynch/stdnal.rs#L88).